include/sparrow/array_data.hpp

```diff
@@ -36,8 +50,251 @@ namespace sparrow
         std::vector<array_data> child_data;
     };
 
-    struct null_type
+    /**
+     * Reference type used in layout classes.
+     * This classes provides a similar behavior
+     * as std::optional, but on references in
+     * an array_data.
+     */
+    template <class L>
+    class reference_proxy
     {
+    public:
+
+        using self_type = reference_proxy<L>;
+        using layout_type = L;
+        using value_type = typename L::inner_value_type;
+        using size_type = typename L::size_type;
+
+        reference_proxy(layout_type& l, size_type index);
+
+        bool has_value() const;
+        explicit operator bool() const;
+
+        value_type& value();
+        const value_type& value() const;
+
+        self_type& operator=(const self_type& rhs);
+        self_type& operator=(self_type&& rhs);
+        self_type& operator=(std::nullopt_t);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(const std::optional<U>& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(std::optional<U>&& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(U&& value);
+
+        void reset();
```

Comment:
It's not clear to me what this does exactly? Needs documentation, like `post-condition: has_value() == false` ?

Reply:
I was confused with `reset` from `std::optional`. I'll move this method to the private section.